worker/config.go

```diff
@@ -0,0 +1,180 @@
+package worker
+
+import (
+	schematypes "github.com/taskcluster/go-schematypes"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+)
+
+type configType struct {
+	Engine          string                 `json:"engine"`
+	Engines         map[string]interface{} `json:"engines"`
+	Plugins         interface{}            `json:"plugins"`
+	Capacity        int                    `json:"capacity"`
+	Credentials     credentials            `json:"credentials"`
+	PollingInterval int                    `json:"pollingInterval"`
+	ReclaimOffset   int                    `json:"reclaimOffset"`
+	QueueBaseURL    string                 `json:"queueBaseUrl"`
+	ProvisionerID   string                 `json:"provisionerId"`
+	WorkerType      string                 `json:"workerType"`
+	WorkerGroup     string                 `json:"workerGroup"`
+	WorkerID        string                 `json:"workerId"`
+}
+
+type credentials struct {
+	ClientID    string `json:"clientId"`
+	AccessToken string `json:"accessToken"`
+	Certificate string `json:"certificate"`
+}
+
+// ConfigSchema returns the configuration schema for the worker.
+func ConfigSchema() schematypes.Object {
+	engineConfig := schematypes.Properties{}
+	engineNames := []string{}
+	for name, provider := range engines.Engines() {
+		engineNames = append(engineNames, name)
+		engineConfig[name] = provider.ConfigSchema()
+	}
+	return schematypes.Object{
+		MetaData: schematypes.MetaData{
+			Title:       "Worker Configuration",
+			Description: `This contains configuration for the worker process.`,
+		},
+		Properties: schematypes.Properties{
+			"engine": schematypes.StringEnum{
+				MetaData: schematypes.MetaData{
+					Title: "Worker Engine",
+					Description: `Selected worker engine to use, notice that the
+						configuration for this engine **must** be present under the
+						'engines.<engine>' configuration key.`,
+				},
+				Options: engineNames,
+			},
+			"engines": schematypes.Object{
+				MetaData: schematypes.MetaData{
+					Title: "Engine Configuration",
+					Description: `Mapping from engine name to engine configuration.
+						Even-though the worker will only use one engine at any given time,
+						the configuration file can hold configuration for all engines.
+						Hence, you need only update the 'engine' key to change which engine
+						should be used.`,
+				},
+				Properties: engineConfig,
```

Comment:
Now isn't this cool... notice that in the type above `Engines` has the type `map[string]interface{}`, so it works even though the engines may specify whatever they want as config --- I can still unmap it into something well-typed.